lib/internal/modules/esm/loader.js

```diff
@@ -151,22 +151,28 @@ class DefaultModuleLoader {
     return this.getJobFromResolveResult(resolveResult, parentURL, importAssertions);
   }
 
-  getJobFromResolveResult(
+  async getJobFromResolveResult(
     resolveResult,
     parentURL,
     importAssertions = resolveResult.importAssertions,
   ) {
     const { url, format } = resolveResult;
-debug('getJobFromResolveResult:', { [url]: { format, moduleMap: this.moduleMap } });
+
+     // `job` may be a Promise, which the engine INSANELY _sometimes_ considers undefined and
+     // sometimes not. This is why we use `has` instead of `get` to check for its existence.
+     // ! Do NOT try to check against its value; the engine will **gladly** screw you over.
+    if (!this.moduleMap.has(url, importAssertions.type)) {
+      debug('getJobFromResolveResult: did NOT find existing entry for %s', url)
+      return this.#createModuleJob(url, parentURL, importAssertions, format);
+    }
+
     let job = this.moduleMap.get(url, importAssertions.type);
+    debug('getJobFromResolveResult: found existing entry for %s %o', url, job);
 
-    // CommonJS will set functions for lazy job evaluation.
-    if (typeof job === 'function') {
+    if (typeof job === 'function') { // CommonJS will set functions for lazy job evaluation.
       this.moduleMap.set(url, undefined, job = job());
```

Comment:
and this line suggests in fact it has a unique non-Map-like API, since `set` also doesn't take more than 2 arguments typically.